### PR TITLE
[SCB-358] fix bug for monitor output id that register only name without any tags

### PR DESCRIPTION
**Pull request by zhengyangyong**
Signed-off-by: zhengyangyong <yangyong.zheng@huawei.com>

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
User can register monitor only name without any tags like :
*MonitorManager.getInstance().getCounter("MonitorWithoutAnyTag");*
In this case output metric id will be **MonitorWithoutAnyTag)** , correct output metric id is **MonitorWithoutAnyTag** ,not include a **')'**

Also fix MetricsLoader load this style of metric id.

### CODE REVIEW

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java [hunk 1]
```diff
@@ -71,11 +95,28 @@ public boolean containsTag(String tagKey, String tagValue) {
...
+    int count = 0;
+    for (char cr : id.toCharArray()) {
+      if (cr == c) {
+        count++;
+      }
+    }
+    return count;
+  }
+
+  private boolean isCorrectId(String id) {
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: what if the id is empty string such as "".

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    //ignore because throw exception is correct
+    catch (ServiceCombException ignore) {
+    }
+  }
+
+  private void checkBadIdFormat(String id) throws Exception {
+    try {
+      new Metric(id, 100);
+      throw new Exception("CheckFailed");
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java]: You can use fail("xxx") here.

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    try {
+      metric.containsTag("A", "1", "B", "X", "C");
+      throw new Exception("CheckFailed");
+    }
+    //ignore because throw exception is correct
+    catch (ServiceCombException ignore) {
+    }
+  }
+
+  private void checkBadIdFormat(String id) throws Exception {
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java]: Using the Expected Exception can do the work. 
https://github.com/junit-team/junit4/wiki/exception-testing

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricsLoader.java [hunk 4]
```diff
@@ -53,4 +54,15 @@ public void checkGetChildrenCount() {
...
+
+  @Test
+  public void checkNoSuchId() throws Exception {
+    try {
+      loader.getMetricTree("Z");
+      throw new Exception("CheckFailed");
+    }
+    //ignore because throw exception is correct
+    catch (ServiceCombException ignore) {
+    }
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricsLoader.java]: We need to verify the ServiceCombException message here.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java [hunk 5]
```diff
@@ -78,43 +82,58 @@ private void setupWindowTime() {
   }
 
   public Counter getCounter(String name, String... tags) {
-    return counters.computeIfAbsent(getMonitorKey(name, tags), f -> {
-      Counter counter = new BasicCounter(getConfig(name, tags));
-      basicMonitorRegistry.register(counter);
-      return counter;
-    });
+    if (isCorrectMonitorNameAndTags(name, tags)) {
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java]: It's more easy to throw the exception here.

File: metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java [hunk 6]
```diff
@@ -24,46 +24,61 @@
...
+    reset();
+    HealthCheckerManager.getInstance().register(good);
+    HealthCheckerPublisher publisher = new HealthCheckerPublisher();
+    Assert.assertEquals(true, publisher.checkHealth());
   }
 
   @Test
-  public void checkHealth() {
+  public void checkHealthBad() {
+    reset();
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java]: Done

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java [hunk 7]
```diff
@@ -182,4 +206,8 @@ private void registerSystemMetrics() {
   private <V extends Number> void registerSystemMetricItem(Callable<V> callable, String name) {
     this.getGauge(callable, MetricsConst.JVM, MetricsConst.TAG_STATISTIC, "gauge", MetricsConst.TAG_NAME, name);
   }
+
+  private boolean isCorrectMonitorNameAndTags(String name, String... tags) {
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java]: Done

File: foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/event/TestEventBus.java [hunk 8]
```diff
@@ -81,4 +81,28 @@ public void process(String data) {
     Thread.sleep(1000);
     Assert.assertFalse(eventReceived.get());
   }
+
+  @Test
+  public void checkUnmatchTypeWillNotReceived() throws InterruptedException {
+    AtomicBoolean eventReceived = new AtomicBoolean(false);
+
+    EventListener<String> listener = new EventListener<String>() {
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/event/TestEventBus.java]: Done

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java [hunk 9]
```diff
@@ -35,14 +36,33 @@ public String getName() {
   }
 
   public Metric(String id, double value) {
-    String[] nameAndTag = id.split("\\(");
-    this.tags = new HashMap<>();
-    String[] tagAnValues = nameAndTag[1].split("[=,)]");
-    for (int i = 0; i < tagAnValues.length; i += 2) {
-      this.tags.put(tagAnValues[i], tagAnValues[i + 1]);
+    if (isCorrectId(id)) {
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: Done

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java [hunk 10]
```diff
@@ -35,14 +36,33 @@ public String getName() {
...
+      this.tags = new HashMap<>();
+      this.value = value;
+      String[] nameAndTag = id.split("[()]");
+      if (nameAndTag.length == 1) {
+        if (!id.endsWith(")")) {
+          this.name = nameAndTag[0];
+        } else {
+          throw new ServiceCombException("bad format id");
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: Done

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java (hunk 1 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: Done

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java (hunk 2 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java]: Done

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java (hunk 3 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetric.java]: Done

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricsLoader.java (hunk 4 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricsLoader.java]: Done

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java [hunk 11]
```diff
@@ -71,11 +95,28 @@ public boolean containsTag(String tagKey, String tagValue) {
...
+    if (tags.length >= 2 && tags.length % 2 == 0) {
+      for (int i = 0; i < tags.length; i += 2) {
+        if (!containsTag(tags[i], tags[i + 1])) {
+          return false;
+        }
       }
+      return true;
     }
-    return true;
+    throw new ServiceCombException("bad tags count");
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: Done

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java [hunk 12]
```diff
@@ -35,14 +36,33 @@ public String getName() {
   }
 
   public Metric(String id, double value) {
-    String[] nameAndTag = id.split("\\(");
-    this.tags = new HashMap<>();
-    String[] tagAnValues = nameAndTag[1].split("[=,)]");
-    for (int i = 0; i < tagAnValues.length; i += 2) {
-      this.tags.put(tagAnValues[i], tagAnValues[i + 1]);
+    if (validateMetricId(id)) {
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: so many if/else...... flat them.

Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: why need to validate id?
servo/spectator's id, we can not use?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java [hunk 13]
```diff
@@ -35,14 +36,33 @@ public String getName() {
...
+      this.tags = new HashMap<>();
+      this.value = value;
+      String[] nameAndTag = id.split("[()]");
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: why need to split tags?
servo/spectator's already have tags information.

if lost these informations caused by our mechanism, then we should change our mechanism.
this is why i DO NOT agree work on current mechanism.

File: foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/event/TestEventBus.java [hunk 14]
```diff
@@ -81,4 +73,14 @@ public void process(String data) {
     Thread.sleep(1000);
```
Comment (wujimin) [inline foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/event/TestEventBus.java]: why sleep?
our case run too fast????????????????????????????

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java (hunk 12 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: Metric、MetricNode和MetricLoader是工具类用于分析从publish获取的metrics结果Map<String,Double>，如果从servo的config获取会更简单一点，现在从String id parse 比如servicecomb.invocation(operation={operationName},role={role},stage={stage},statistic={statistic},status={status},unit={unit})也是有必要的，我会在新的PR增加你说的内容

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java (hunk 13 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: 我会在新的PR增加从MonitorConfig创建Metric，不必parse

File: foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/event/TestEventBus.java (hunk 14 above)
Comment (zhengyangyong) [reply foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/event/TestEventBus.java]: Fixed,removed

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java (hunk 12 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: Done

Comment (wujimin) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: i will remove metric.java in new PR.

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java (hunk 13 above)
Comment (wujimin) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/Metric.java]: i will remove metric.java in new PR.

**Pull request merged**